### PR TITLE
dapp-test library support

### PR DESCRIPTION
**Pull request by rainbreak**
Depends on https://github.com/dapphub/ethrun/pull/8

Todo:
- [x] fix `ethrun` falling over on multiple contracts
- ~~what happens for `dapp create`?~~ [out of scope]
- ~~allow some kind of link spec in dappfile?~~[out of scope]

### CODE REVIEW

**Comment by rainbreak** [comment]
Could leave the `dapp create` / dappfile as out of scope for this PR.

File: libexec/dapp/dapp-test [hunk 1]
```diff
@@ -1,4 +1,24 @@
...
+libraries=$(grep -EoR '^library\s+\w+' $DAPP_SRC $DAPP_LIB \
+          | sed -E 's/:library\s+/:/' \
+          | sort)
+links=""
+for lib in $libraries; do
+    address="$(dapp address $sender $nonce)"
+    links="$links $lib:$address"
+    ((nonce++))
+done
+(set -x; solc --link --libraries "$links" $DAPP_OUT/*.bin)
```
Comment (mhhf) [inline libexec/dapp/dapp-test]: This won't work on OS X with `solc 0.4.12` and BSD sed
with `$libraries = "src/lib.sol:library Lib"` (testenv)
and `$deploy = "out/library Lib.bin"`
It feels like sed is working differently on os x -.-, will investigate

Comment (rainbreak) [inline libexec/dapp/dapp-test]: Thanks. I did what I thought was cross platform, but have no idea about osx really.

Comment (rainbreak) [inline libexec/dapp/dapp-test]: I think `grep` is the problem. `-R` and `-o` aren't in POSIX, but it looks like `-o` is supported on osx.

Comment (rainbreak) [inline libexec/dapp/dapp-test]: @mhhf try 32c4fc5. I'll remove the `-o` as well if needed.

File: libexec/dapp/dapp-address [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+// shut up secp256k1
+const log = console.log
+console.log = () => {}
+const generateAddress = require("ethereumjs-util").generateAddress
+console.log = log
+
+var sender = PROGRAM_ARGS[0]
+var nonce = PROGRAM_ARGS[1]
+
+console.log(generateAddress(sender, nonce).toString("hex"))
```
Comment (rainbreak) [inline libexec/dapp/dapp-address]: @dbrock: I think this should maybe be in `seth` instead.

Can also use with `seth nonce` to determine the address of the next contract to be generated by an address.

File: libexec/dapp/dapp-test [hunk 3]
```diff
@@ -8,8 +8,8 @@ nonce=1048576  # magic starting nonce
 
 # need to know the names and locations of all of the libraries
 sources=$(find $DAPP_SRC $DAPP_LIB -name '*.sol')
-libraries=$(grep -Eo '^library\s+\w+' $sources \
-          | sed -E 's/:library\s+/:/' \
+libraries=$(grep -E '^library\s+\w+' $sources \
+          | sed -E 's/:library\s+(\w+)\s+\{/:\1/' \
```
Comment (rainbreak) [inline libexec/dapp/dapp-test]: Apparently backreferences aren't POSIX, but osx supports them. ¯\_(ツ)_/¯

**Pull request merged**